File: vscode-extension/scripts/loop_detector_ast.py
import ast, json, sys

def get_call_name(node):
    try:
        if isinstance(node, ast.Name): return node.id
        if isinstance(node, ast.Attribute):
            p = get_call_name(node.value)
            return (p + '.' + node.attr) if p else node.attr
    except Exception: pass
    return ''

COST_PATTERNS = [
    'completions.create', 'messages.create', 'generatecontent', 'generatetext', 'streamtext',
    'chatcompletion.create', 'completion.create',
    'axios.get', 'axios.post', 'axios.put', 'axios.delete', 'fetch',
    'requests.get', 'requests.post', 'requests.put', 'requests.delete',
    'httpx.get', 'httpx.post', 'httpx.put', 'httpx.delete',
    'prisma.', 'findunique', 'findmany', 'findbyid', 'findone',
    'dynamodb', 'docclient', 'table.scan', 'client.scan',
]

def is_costly(name): return any(p in name.lower() for p in COST_PATTERNS)

class Visitor(ast.NodeVisitor):
    def __init__(self):
        self.loop_line = None
        self.hits = []

    def _loop(self, node):
        prev = self.loop_line
        self.loop_line = node.lineno
        self.generic_visit(node)
        self.loop_line = prev

    visit_For = visit_While = visit_AsyncFor = _loop

    def visit_Call(self, node):
        if self.loop_line is not None:
            name = get_call_name(node.func)
            if name and is_costly(name):
                self.hits.append({
                    'line': node.lineno,
                    'col': node.col_offset,
                    'callName': name,
                    'loopStartLine': self.loop_line,
                })
        self.generic_visit(node)

fp = sys.argv[1]
try:
    src = open(fp, encoding='utf-8', errors='replace').read()
    v = Visitor()
    v.visit(ast.parse(src, filename=fp))
    print(json.dumps(v.hits))
except SyntaxError:
    print(json.dumps([]))

File: vscode-extension/scripts/loop_detector_ast.go
package main

import (
	"encoding/json"
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"strings"
)

var costPatterns = []string{
	"completions.new", "messages.new",        // openai-go, anthropic sdk-go
	"createchatcompletion", "createcompletion", // go-openai
	"generatecontent", "generatetext",
	"http.get", "http.post", "http.put", "http.delete",
	"client.get", "client.post", "client.put", "client.delete",
	"do(", // http.Client.Do
	"findone", "findmany", "find(",
	"table.scan", "client.scan",
}

func isCostly(name string) bool {
	lower := strings.ToLower(name)
	for _, p := range costPatterns {
		if strings.Contains(lower, p) {
			return true
		}
	}
	return false
}

func callName(expr ast.Expr) string {
	switch e := expr.(type) {
	case *ast.Ident:
		return e.Name
	case *ast.SelectorExpr:
		parent := callName(e.X)
		if parent == "" {
			return e.Sel.Name
		}
		return parent + "." + e.Sel.Name
	}
	return ""
}

type Hit struct {
	Line          int    `json:"line"`
	Col           int    `json:"col"`
	CallName      string `json:"callName"`
	LoopStartLine int    `json:"loopStartLine"`
}

func walk(node ast.Node, fset *token.FileSet, loopLine int, hits *[]Hit) {
	if node == nil {
		return
	}
	switch n := node.(type) {
	case *ast.ForStmt, *ast.RangeStmt:
		pos := fset.Position(node.Pos())
		ast.Inspect(node, func(inner ast.Node) bool {
			if inner == node {
				return true
			}
			if call, ok := inner.(*ast.CallExpr); ok {
				name := callName(call.Fun)
				if name != "" && isCostly(name) {
					p := fset.Position(call.Pos())
					*hits = append(*hits, Hit{
						Line:          p.Line,
						Col:           p.Column - 1,
						CallName:      name,
						LoopStartLine: pos.Line,
					})
				}
			}
			return true
		})
		return // already walked inside
	case *ast.File:
		for _, decl := range n.Decls {
			walk(decl, fset, loopLine, hits)
		}
	case *ast.FuncDecl:
		if n.Body != nil {
			walk(n.Body, fset, loopLine, hits)
		}
	case *ast.BlockStmt:
		for _, stmt := range n.List {
			walk(stmt, fset, loopLine, hits)
		}
	default:
		ast.Inspect(n, func(inner ast.Node) bool {
			if inner == n {
				return true
			}
			if _, ok := inner.(*ast.ForStmt); ok {
				walk(inner, fset, loopLine, hits)
				return false
			}
			if _, ok := inner.(*ast.RangeStmt); ok {
				walk(inner, fset, loopLine, hits)
				return false
			}
			return true
		})
	}
}

func main() {
	if len(os.Args) < 2 {
		json.NewEncoder(os.Stdout).Encode([]Hit{})
		return
	}

	fp := os.Args[1]
	fset := token.NewFileSet()
	f, err := parser.ParseFile(fset, fp, nil, 0)
	if err != nil {
		json.NewEncoder(os.Stdout).Encode([]Hit{})
		return
	}

	var hits []Hit
	walk(f, fset, 0, &hits)
	if hits == nil {
		hits = []Hit{}
	}
	json.NewEncoder(os.Stdout).Encode(hits)
}
